File: coordinator/internal/e2e/e2e.go
// Package e2e implements the end-to-end encryption between the gateway and
// provider daemons:
//
//	X25519 ephemeral↔static ECDH → HKDF-SHA256 → XChaCha20-Poly1305
//	(+ Ed25519 signatures for engine-reported usage)
//
// Properties: the coordinator relays prompt ciphertext without being able to
// read it; each request uses a fresh ephemeral key (forward secrecy); each
// response leg uses a gateway-generated response key. This is the modern
// equivalent of Darkbloom's NaCl Box scheme (XSalsa20 is unavailable in
// Apple's Swift Crypto, ChaChaPoly is native in both stacks).
package e2e

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"fmt"

	"golang.org/x/crypto/chacha20poly1305"
	"golang.org/x/crypto/curve25519"
	"golang.org/x/crypto/hkdf"
)

const (
	// KeySize is the X25519/Ed25519 raw key size (32 bytes).
	KeySize = 32
	// NonceSize is the ChaCha20-Poly1305 nonce size (12 bytes; fresh ECDH
	// per payload makes random nonces safe).
	NonceSize = 12
	// Info is the HKDF info string binding keys to this protocol.
	Info = "idlegrid-e2e-v1"
)

// XKeyPair is an X25519 key pair (raw 32-byte keys).
type XKeyPair struct {
	Priv, Pub [32]byte
}

// GenerateX25519 creates a fresh X25519 key pair.
func GenerateX25519() XKeyPair {
	var kp XKeyPair
	if _, err := rand.Read(kp.Priv[:]); err != nil {
		panic(err)
	}
	pub, err := curve25519.X25519(kp.Priv[:], curve25519.Basepoint)
	if err != nil {
		panic(err)
	}
	copy(kp.Pub[:], pub)
	return kp
}

// EdKeyPair is an Ed25519 signing key pair.
type EdKeyPair struct {
	Pub  ed25519.PublicKey
	Priv ed25519.PrivateKey
}

// GenerateEd25519 creates a fresh signing key pair.
func GenerateEd25519() EdKeyPair {
	_, priv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		panic(err)
	}
	return EdKeyPair{Pub: priv.Public().(ed25519.PublicKey), Priv: priv}
}

func deriveKey(shared, salt []byte) ([]byte, error) {
	r := hkdf.New(sha256.New, shared, salt, []byte(Info))
	key := make([]byte, chacha20poly1305.KeySize)
	if _, err := r.Read(key); err != nil {
		return nil, err
	}
	return key, nil
}

// Seal encrypts plaintext to the recipient's X25519 public key using a fresh
// ephemeral sender key. Returns the base64 eph pubkey, nonce, and ciphertext.
func Seal(plaintext []byte, recipientPub *[32]byte) (ephPubB64, nonceB64, ctB64 string, err error) {
	eph := GenerateX25519()
	shared, err := curve25519.X25519(eph.Priv[:], recipientPub[:])
	if err != nil {
		return "", "", "", err
	}
	nonce := make([]byte, NonceSize)
	if _, err := rand.Read(nonce); err != nil {
		return "", "", "", err
	}
	key, err := deriveKey(shared, nonce)
	if err != nil {
		return "", "", "", err
	}
	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return "", "", "", err
	}
	ct := aead.Seal(nil, nonce, plaintext, nil)
	return base64.StdEncoding.EncodeToString(eph.Pub[:]),
		base64.StdEncoding.EncodeToString(nonce),
		base64.StdEncoding.EncodeToString(ct), nil
}

// Open decrypts a sealed payload with the recipient's private key.
func Open(ephPubB64, nonceB64, ctB64 string, recipientPriv *[32]byte) ([]byte, error) {
	ephPub, err := base64.StdEncoding.DecodeString(ephPubB64)
	if err != nil || len(ephPub) != KeySize {
		return nil, fmt.Errorf("bad eph pubkey")
	}
	nonce, err := base64.StdEncoding.DecodeString(nonceB64)
	if err != nil || len(nonce) != NonceSize {
		return nil, fmt.Errorf("bad nonce")
	}
	ct, err := base64.StdEncoding.DecodeString(ctB64)
	if err != nil {
		return nil, fmt.Errorf("bad ciphertext")
	}
	shared, err := curve25519.X25519(recipientPriv[:], ephPub)
	if err != nil {
		return nil, err
	}
	key, err := deriveKey(shared, nonce)
	if err != nil {
		return nil, err
	}
	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return nil, err
	}
	return aead.Open(nil, nonce, ct, nil)
}

// Sign returns the base64 Ed25519 signature of message.
func Sign(k EdKeyPair, message []byte) string {
	return base64.StdEncoding.EncodeToString(ed25519.Sign(k.Priv, message))
}

// Verify checks a base64 Ed25519 signature over message.
func Verify(pubB64 string, message []byte, sigB64 string) bool {
	pub, err := base64.StdEncoding.DecodeString(pubB64)
	if err != nil || len(pub) != ed25519.PublicKeySize {
		return false
	}
	sig, err := base64.StdEncoding.DecodeString(sigB64)
	if err != nil || len(sig) != ed25519.SignatureSize {
		return false
	}
	return ed25519.Verify(ed25519.PublicKey(pub), message, sig)
}

// PubKeyFromB64 parses a base64 X25519 public key.
func PubKeyFromB64(b64 string) (*[32]byte, error) {
	raw, err := base64.StdEncoding.DecodeString(b64)
	if err != nil || len(raw) != KeySize {
		return nil, fmt.Errorf("bad public key")
	}
	var out [32]byte
	copy(out[:], raw)
	return &out, nil
}

// PubKeyToB64 encodes an X25519 public key.
func PubKeyToB64(pub *[32]byte) string {
	return base64.StdEncoding.EncodeToString(pub[:])
}

File: provider-swift/Sources/idlegrid-provider/E2E.swift
import Crypto
import Foundation

/// E2E crypto identity for the provider daemon.
///
/// Request leg:  gateway ephemeral X25519 ↔ provider static X25519 → HKDF-SHA256 → ChaChaPoly
/// Response leg: provider ephemeral X25519 ↔ gateway response key
/// Usage auth:   Ed25519 signature over the usage JSON (provider's signing key)
///
/// Matches coordinator/internal/e2e (Go) byte-for-byte.
struct E2EIdentity {
    let x25519: Curve25519.KeyAgreement.PrivateKey
    let signing: Curve25519.Signing.PrivateKey

    init() {
        x25519 = Curve25519.KeyAgreement.PrivateKey()
        signing = Curve25519.Signing.PrivateKey()
    }

    var publicKeyB64: String { Data(x25519.publicKey.rawRepresentation).base64EncodedString() }
    var signingKeyB64: String { Data(signing.publicKey.rawRepresentation).base64EncodedString() }
}

enum E2E {
    static let info = Data("idlegrid-e2e-v1".utf8)

    static func b64(_ data: Data) -> String { data.base64EncodedString() }
    static func unb64(_ s: String) -> Data? { Data(base64Encoded: s) }

    private static func deriveKey(_ shared: SharedSecret, salt: Data) -> SymmetricKey {
        shared.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: salt,
            sharedInfo: info,
            outputByteCount: 32
        )
    }

    /// Seals plaintext to the peer's X25519 public key with a fresh ephemeral
    /// key + random nonce. Returns the SealedPayload fields (base64).
    static func seal(_ plaintext: Data, peerPubB64: String,
                     identity: E2EIdentity) throws -> (ephPub: String, nonce: String, ciphertext: String) {
        try sealWithEphemeral(plaintext, peerPubB64: peerPubB64, eph: identity.x25519)
    }

    /// Seals with an explicit ephemeral key (response leg: one per request).
    static func sealWithEphemeral(_ plaintext: Data, peerPubB64: String,
                                  eph: Curve25519.KeyAgreement.PrivateKey) throws -> (ephPub: String, nonce: String, ciphertext: String) {
        guard let peerRaw = unb64(peerPubB64), peerRaw.count == 32 else {
            throw E2EError.badPeerKey
        }
        let peer = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: peerRaw)
        let shared = try eph.sharedSecretFromKeyAgreement(with: peer)
        let nonceData = (0..<12).map { _ in UInt8.random(in: 0...255) }
        let key = deriveKey(shared, salt: Data(nonceData))
        let sealed = try ChaChaPoly.seal(plaintext, using: key, nonce: ChaChaPoly.Nonce(data: Data(nonceData)))
        // Go's AEAD output is ciphertext||tag — concatenate both.
        let ctWithTag = sealed.ciphertext + sealed.tag
        return (b64(Data(eph.publicKey.rawRepresentation)), b64(Data(nonceData)), b64(ctWithTag))
    }

    /// Opens a payload sealed to our static key by the peer's ephemeral key.
    static func open(ephPubB64: String, nonceB64: String, ctB64: String,
                     identity: E2EIdentity) throws -> Data {
        guard let eph = unb64(ephPubB64), eph.count == 32,
              let nonce = unb64(nonceB64), nonce.count == 12,
              let ct = unb64(ctB64), ct.count > 16
        else { throw E2EError.badPayload }
        let peer = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: eph)
        let shared = try identity.x25519.sharedSecretFromKeyAgreement(with: peer)
        let key = deriveKey(shared, salt: nonce)
        // Go sends nonce separately; AEAD output is ciphertext||tag.
        // Reconstruct CryptoKit's combined format: nonce || ciphertext || tag.
        let sealedBox = try ChaChaPoly.SealedBox(combined: nonce + ct)
        return try ChaChaPoly.open(sealedBox, using: key)
    }

    /// Signs usage JSON with the provider's Ed25519 key.
    static func sign(_ data: Data, identity: E2EIdentity) -> String? {
        (try? identity.signing.signature(for: data))?.base64EncodedString()
    }

    /// Verifies a base64 Ed25519 signature over data.
    static func verify(pubB64: String, data: Data, sigB64: String) -> Bool {
        guard let pub = unb64(pubB64), pub.count == 32,
              let sig = unb64(sigB64) else { return false }
        return identityVerify(pub, data, sig)
    }

    private static func identityVerify(_ pub: Data, _ data: Data, _ sig: Data) -> Bool {
        guard let key = try? Curve25519.Signing.PublicKey(rawRepresentation: pub) else { return false }
        return key.isValidSignature(sig, for: data)
    }
}

enum E2EError: Error {
    case badPeerKey
    case badPayload
    case decryptFailed
}
